ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/FutureAssertions.scala

```diff
@@ -34,13 +40,126 @@ final class FutureAssertions[T](future: Future[T]) {
       case Success(value) => Failure(new ExpectedFailureException(context, value))
       case Failure(other) => Failure(other)
     }
+
 }
 
 object FutureAssertions {
 
+  def assertAfter[V](
+      delay: FiniteDuration
+  )(test: => Future[V]): Future[V] = {
+    Delayed.Future.by(delay)(test)
+  }
+
+  def succeedsEventually[V](
+      delay: FiniteDuration = 100.millis,
+      maxInterval: FiniteDuration,
+      description: String,
+  )(
+      test: => Future[V]
+  )(implicit ec: ExecutionContext, logger: Logger): Future[V] = {
+    def internalSucceedsEventually(interval: FiniteDuration): Future[V] = {
+      val newMaxInterval = interval - delay
+      if (newMaxInterval < Duration.Zero) {
+        test.andThen { case Failure(exception) =>
+          logger.error(
+            s"Assertion never succeeded after $maxInterval with a delay of $delay. Description: $description",
+            exception,
+          )
+        }
+      } else {
+        if (newMaxInterval == Duration.Zero) {
+          assertAfter(delay)(test)
+        } else {
+          assertAfter(delay)(test).recoverWith { case NonFatal(ex) =>
+            logger.debug(
+              s"Failed assertion: $description. Running again with new max interval $newMaxInterval",
+              ex,
+            )
+            internalSucceedsEventually(newMaxInterval)
+          }
+        }
+      }
+    }
+
+    internalSucceedsEventually(maxInterval)
+  }
+
+  def succeedsUntil[V](
+      delay: FiniteDuration = 100.millis,
+      succeedDuration: FiniteDuration,
+      succeedDeadline: Option[Instant] = None,
+  )(
+      test: => Future[V]
+  )(implicit ec: ExecutionContext, logger: Logger): Future[V] = {
```

Comment:
Loggers are not designed to be passed around; you have one per class. If you need context, use `ContextualizedLogger` and `LoggingContext`.